File: js/binary_search.js
var arguments = process.argv;

if (arguments.length < 4) {
  console.log('usage: binary_search.js [-h] item array [array ...]');
  return;
}

let item = parseInt(arguments[2]);
let array = [];

for (var i = 3; i < arguments.length; i++) {
  array.push(parseInt(arguments[i]));
}

if (binary_search(array, item)) {
  console.log('Found it!');
} else {
  console.log('Not found...');
}

function binary_search(array, item) {
  console.log('Searching for ' + item + ' in [' + array + ']');
  let left = 0;
  let right = array.length - 1;

  while (left <= right) {
    let middle = Math.floor((left + right) / 2);

    if (array[middle] < item) {
      left = middle + 1;
    } else if (array[middle] > item) {
      right = middle - 1;
    } else {
      return true;
    }
  }

  return false;
}

File: c/binary_search.c
#include "common.h"

int binary_search(int value, int array[], int length)
{
    printf("Searching for %d in ", value);
    print_array(array, length);
    printf("\n");

    int left = 0;
    int right = length - 1;

    while (left <= right)
    {
        int middle = (int)((left + right) / 2);
        if (array[middle] < value)
        {
            left = middle + 1;
        }
        else if (array[middle] > value)
        {
            right = middle - 1;
        }
        else
        {
            return 1;
        }
    }

    return 0;
}

int main(int argc, char **argv)
{
    println("Running tests...");

    int array[] = {1, 2, 3, 4, 5};
    int length = lengthof(array);
    int empty_array[] = {};

    assert(binary_search(2, array, length) == 1);
    assert(binary_search(10, array, length) == 0);
    assert(binary_search(1, empty_array, 0) == 0);

    println("All tests passed!");
}